### PR TITLE
Add Rector

### PR DESCRIPTION
**Pull request by MGatner**

### CODE REVIEW

**Comment by MGatner** [comment]
@samsonasik See what you think of this. Based on the version from the CI4 repo with the "utility" classes and file references removed. Once PHP 7.4 is the minimum we can update some of these rule sets to be more inclusive.

File: src/.github/workflows/rector.yml [hunk 1]
```diff
@@ -0,0 +1,68 @@
...
+          composer -q config -g github-oauth.github.com "${{ secrets.GITHUB_TOKEN }}"
+          if [ -f composer.lock ]; then
+            composer install --no-progress --no-interaction --prefer-dist --optimize-autoloader
+          else
+            composer update --no-progress --no-interaction --prefer-dist --optimize-autoloader
+          fi
+
+      - name: Analyze for refactoring
+        run: |
+          composer global require --dev rector/rector
```
Comment (samsonasik) [inline src/.github/workflows/rector.yml:67]: specific version seems needed: rector/rector:0.12.4

File: src/rector.php [hunk 2]
```diff
@@ -0,0 +1,115 @@
...
+    $containerConfigurator->import(PHPUnitSetList::PHPUNIT_80);
+
+    $parameters = $containerConfigurator->parameters();
+
+    // paths to refactor; solid alternative to CLI arguments
+    $parameters->set(Option::PATHS, [__DIR__ . '/app', __DIR__ . '/tests']);
+
+    // do you need to include constants, class aliases or custom autoloader? files listed will be executed
+    $parameters->set(Option::BOOTSTRAP_FILES, [
+        realpath(getcwd()) . '/vendor/autoload.php',
```
Comment (samsonasik) [inline src/rector.php:53]: `vendor/autoload.php` can be in `Option::AUTOLOAD_PATHS`, ref https://github.com/rectorphp/rector/blob/main/docs/static_reflection_and_autoload.md#static-reflection-and-autoload

File: src/.github/workflows/rector.yml [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+          composer -q config -g github-oauth.github.com "${{ secrets.GITHUB_TOKEN }}"
+          if [ -f composer.lock ]; then
+            composer install --no-progress --no-interaction --prefer-dist --optimize-autoloader
+          else
+            composer update --no-progress --no-interaction --prefer-dist --optimize-autoloader
+          fi
+
+      - name: Analyze for refactoring
+        run: |
+          composer global require --dev rector/rector:^0.12.4
```
Comment (samsonasik) [inline src/.github/workflows/rector.yml:67]: better to have pinned version so we can verify every version to verify bug fix, like in https://github.com/codeigniter4/CodeIgniter4/pull/5328#issuecomment-969483172

```suggestion
          composer global require --dev rector/rector:0.12.4
```

File: src/rector.php (hunk 2 above)
Comment (MGatner) [reply src/rector.php:53]: That actually makes me think that we don't need to reference the autoload file at all, right?

File: src/.github/workflows/rector.yml (hunk 3 above)
Comment (MGatner) [reply src/.github/workflows/rector.yml:67]: I'm not sure how to handle this one. The goal is for these to be templates that developers can merge into their own repos and then adjust as necessary. I don't want us to have to keep updating this repo every time there is a new minor patch to Rector. Maybe we reference the minor version `^0.12` and add a note to the docs that developers should consider locking it to the specific patch?

Comment (MGatner) [reply src/.github/workflows/rector.yml]: I rolled it back to the unspecified version and added a note in the docs; see what you think.

File: src/rector.php (hunk 2 above)
Comment (MGatner) [reply src/rector.php]: I removed the Composer path altogether (it seems like Rector finds this on its own) and cleaned the config file up a bunch. Re-review please?

File: src/rector.php [hunk 4]
```diff
@@ -27,69 +27,62 @@
...
     $containerConfigurator->import(PHPUnitSetList::PHPUNIT_80);
 
     $parameters = $containerConfigurator->parameters();
 
-    // paths to refactor; solid alternative to CLI arguments
-    $parameters->set(Option::PATHS, [__DIR__ . '/app', __DIR__ . '/tests']);
+    // The paths to refactor (can also be supplied with CLI arguments)
+    $parameters->set(Option::PATHS, [
+        __DIR__ . '/app',
+        __DIR__ . '/tests',
```
Comment (samsonasik) [inline src/rector.php:49]: is this not used in CodeIgniter4 project as not check `system` dir?

File: README.md [hunk 5]
```diff
@@ -130,6 +130,9 @@ predefined rulesets. Rector can be highly opinionated based on its configuration
 so be sure to read the documentation and figure out the best fit for you. This workflow performs
 a "dry run" to check for any changes that Rector would have made and fail if there are matches.
 
+> Note: Rector updates rules all the time so you may want to lock your repo to the latest known working version of Rector to prevent unexpected failures
+> E.g. in **.github/workflows/rector.yml** supply the specific minor patch: `composer global require --dev rector/rector:0.12.4`
+
```
Comment (samsonasik) [inline README.md:135]: 👍

**Review by samsonasik — APPROVED** [review]

**Review by kenjis — APPROVED** [review]

File: src/rector.php (hunk 4 above)
Comment (MGatner) [reply src/rector.php:49]: The intent is not for this to be used in the framework repo but distributable to developer's projects and libraries. Libraries will need to adjust the path from **app/** to **src/** (or whatever they use), but that is the case for all the files in this repo and I hope to make a scripted solution down the road.

**Pull request merged**